### PR TITLE
Add method to rationalize `Rational`

### LINKED ISSUE #43369: "`rationalize`" a rational
It might be useful to have `rationalize(x)` return `x` if `x` is already rational, to allow generic code that takes both reals and rationals as inputs.

### PR DESCRIPTION
**Pull request by tantheta01**
Closes #43369.

### CODE REVIEW

**Comment by dkarrasch** [comment]
@tantheta01 Thank you for your contribution and welcome to the community. I helped slightly by (i) making the title more descriptive, and (ii) rewording the OP to the effect that github associates this PR with the issue so when this PR is merged the issue is automatically closed.

File: base/rational.jl [hunk 1]
```diff
@@ -216,8 +216,16 @@ function rationalize(::Type{T}, x::AbstractFloat, tol::Real) where T<:Integer
 end
 rationalize(::Type{T}, x::AbstractFloat; tol::Real = eps(x)) where {T<:Integer} = rationalize(T, x, tol)::Rational{T}
 rationalize(x::AbstractFloat; kvs...) = rationalize(Int, x; kvs...)
-rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re, kvs...)::Rational{T}, rationalize(T, x.im, kvs...)::Rational{T})
```
Comment (JeffBezanson) [inline base/rational.jl:219]: Oh wow, this looks like it was actually a bug. Thanks for catching that. Would be good to add a test for it.

File: base/rational.jl [hunk 2]
```diff
@@ -158,6 +158,7 @@ function rationalize(::Type{T}, x::AbstractFloat, tol::Real) where T<:Integer
     if tol < 0
         throw(ArgumentError("negative tolerance $tol"))
     end
+
```
Comment (JeffBezanson) [inline base/rational.jl:161]: Better to avoid spurious whitespace changes like this.

File: base/rational.jl [hunk 3]
```diff
@@ -216,8 +216,16 @@ function rationalize(::Type{T}, x::AbstractFloat, tol::Real) where T<:Integer
...
-rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re, kvs...)::Rational{T}, rationalize(T, x.im, kvs...)::Rational{T})
-rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re, kvs...), rationalize(Int, x.im, kvs...))
+rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re; kvs...)::Rational{T}, rationalize(T, x.im; kvs...)::Rational{T})
+rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re; kvs...), rationalize(Int, x.im; kvs...))
+rationalize(::Type{T}, x::Rational, tol::Real) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(::Type{T}, x::Rational; tol::Real = zero(x)) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(x::Rational; kvs...) = rationalize(Int, x; kvs...)
+rationalize(::Type{T}, x::Integer, tol::Real) where {T<:Integer} = rationalize(T, Rational(x, 1), tol)
+rationalize(::Type{T}, x::Integer; tol::Real = 0) where {T<:Integer} = Rational(x, 1)
+rationalize(x::Integer; kvs...) = Rational(x, 1)
```
Comment (JeffBezanson) [inline base/rational.jl:226]:
```suggestion
rationalize(x::Integer; kvs...) = Rational(x)
```

File: base/rational.jl [hunk 4]
```diff
@@ -216,8 +216,16 @@ function rationalize(::Type{T}, x::AbstractFloat, tol::Real) where T<:Integer
...
 rationalize(::Type{T}, x::AbstractFloat; tol::Real = eps(x)) where {T<:Integer} = rationalize(T, x, tol)::Rational{T}
 rationalize(x::AbstractFloat; kvs...) = rationalize(Int, x; kvs...)
-rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re, kvs...)::Rational{T}, rationalize(T, x.im, kvs...)::Rational{T})
-rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re, kvs...), rationalize(Int, x.im, kvs...))
+rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re; kvs...)::Rational{T}, rationalize(T, x.im; kvs...)::Rational{T})
+rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re; kvs...), rationalize(Int, x.im; kvs...))
+rationalize(::Type{T}, x::Rational, tol::Real) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(::Type{T}, x::Rational; tol::Real = zero(x)) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(x::Rational; kvs...) = rationalize(Int, x; kvs...)
+rationalize(::Type{T}, x::Integer, tol::Real) where {T<:Integer} = rationalize(T, Rational(x, 1), tol)
```
Comment (JeffBezanson) [inline base/rational.jl:224]:
```suggestion
rationalize(::Type{T}, x::Integer, tol::Real) where {T<:Integer} = Rational{T}(x)
```

File: base/rational.jl [hunk 5]
```diff
@@ -216,8 +216,16 @@ function rationalize(::Type{T}, x::AbstractFloat, tol::Real) where T<:Integer
 end
 rationalize(::Type{T}, x::AbstractFloat; tol::Real = eps(x)) where {T<:Integer} = rationalize(T, x, tol)::Rational{T}
 rationalize(x::AbstractFloat; kvs...) = rationalize(Int, x; kvs...)
-rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re, kvs...)::Rational{T}, rationalize(T, x.im, kvs...)::Rational{T})
-rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re, kvs...), rationalize(Int, x.im, kvs...))
+rationalize(::Type{T}, x::Complex; kvs...) where {T<:Integer} = Complex(rationalize(T, x.re; kvs...)::Rational{T}, rationalize(T, x.im; kvs...)::Rational{T})
+rationalize(x::Complex; kvs...) = Complex(rationalize(Int, x.re; kvs...), rationalize(Int, x.im; kvs...))
+rationalize(::Type{T}, x::Rational, tol::Real) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(::Type{T}, x::Rational; tol::Real = zero(x)) where {T<:Integer} = rationalize(T, float(x), tol)::Rational{T}
+rationalize(x::Rational; kvs...) = rationalize(Int, x; kvs...)
```
Comment (JeffBezanson) [inline base/rational.jl:223]:
```suggestion
rationalize(x::Rational{T}; kvs...) where{T} = rationalize(T, x; kvs...)
```

Comment (JeffBezanson) [reply base/rational.jl:223]: But I wonder if all the cases of `rationalize(x::Rational)` should just return `x`, i.e,. ignore the tolerance argument.

File: base/rational.jl (hunk 1 above)
Comment (inkydragon) [reply base/rational.jl:219]: Maybe something like:

```diff
diff --git a/test/rational.jl b/test/rational.jl
--- test/rational.jl
+++ test/rational.jl
@@ -638,5 +638,11 @@
     @test rationalize(float(pi)im) == 0//1 + 165707065//52746197*im
     @test rationalize(Int8, float(pi)im) == 0//1 + 22//7*im
     @test rationalize(1.192 + 2.233im) == 149//125 + 2233//1000*im
     @test rationalize(Int8, 1.192 + 2.233im) == 118//99 + 67//30*im
+
+    # test: rationalize(x::Complex; kvs...)
+    precise_next = 7205759403792795//72057594037927936
+    @assert Float64(precise_next) == nextfloat(0.1)
+    @test rationalize(nextfloat(0.1) * im; tol=0) == precise_next * im
+    @test rationalize(0.1im; tol=eps(0.1)) == rationalize(0.1im)
 end
```

File: base/rational.jl (hunk 5 above)
Comment (tantheta01) [reply base/rational.jl:223]: I agree on ignoring the tolerance part. However, I feel returning `x` might not work if someone wanted, for instance `rationalize(Int8, 10000/100001)`.

Comment (JeffBezanson) [reply base/rational.jl]: True

**Comment by oscardssmith** [comment]
bumping this. I've added some more tests and I think it's ready to merge.

**Comment by oscardssmith** [comment]
After an embarrassingly long time, this is finally passing tests. merging in a few days sans objections.

**Comment by Keno** [comment]
This failed on CI and shouldn't have been merged. Please fix or revert in the morning.

**Comment by KristofferC** [comment]
Explicitly:

```
Error in testset rational:
--
Test Failed at C:\buildkite-agent\builds\win2k22-amdci6-4\julialang\julia-master\julia-d12ccd04fc\share\julia\test\rational.jl:739
Expression: rationalize(nextfloat(0.1) * im; tol = 0) == precise_next * im
Evaluated: 0//1 + 1//10*im == 0//1 + 7205759403792795//72057594037927936*im
```

**Comment by oscardssmith** [comment]
Sorry about that, I saw that only windows was failing and assumed it was a platform problem. https://github.com/JuliaLang/julia/pull/50163 fixes it (there was an incorrect assumption in the test that `Int` is big enough to store the 53 bits of a Float64 mantissa).

**Pull request merged**